### ISSUE TITLE
ServiceBrowser IndexError in read_name

### ISSUE THREAD

**Issue by gbiddison** [jstasiak/python-zeroconf#63]
Exception while reading name from incoming packet.
Appears to be failure to bounds-check.  

```python
Exception in thread zeroconf-Engine:
Traceback (most recent call last):
  File "//anaconda/envs/py34/lib/python3.4/threading.py", line 911, in _bootstrap_inner
    self.run()
  File "/Users/gbiddison/source/sygo/tools/BeagleDiff/zeroconf.py", line 953, in run
    reader.handle_read(socket_)
  File "/Users/gbiddison/source/sygo/tools/BeagleDiff/zeroconf.py", line 990, in handle_read
    msg = DNSIncoming(data)
  File "/Users/gbiddison/source/sygo/tools/BeagleDiff/zeroconf.py", line 565, in __init__
    self.read_others()
  File "/Users/gbiddison/source/sygo/tools/BeagleDiff/zeroconf.py", line 622, in read_others
    domain, type_, class_, ttl, self.read_name())
  File "/Users/gbiddison/source/sygo/tools/BeagleDiff/zeroconf.py", line 667, in read_name
    length = indexbytes(self.data, off)
IndexError: index out of range
```

Proposed patch:

```diff
diff --git a/zeroconf.py b/zeroconf.py
index 789fc2d..e86fbdb 100644
--- a/zeroconf.py
+++ b/zeroconf.py
@@ -663,10 +663,10 @@ class DNSIncoming(object):
         next_ = -1
         first = off
 
-        while True:
+        while off < len(self.data):
             length = indexbytes(self.data, off)
             off += 1
-            if length == 0:
+            if length == 0 or off >= len(self.data):
                 break
             t = length & 0xC0
             if t == 0x00:
```

**Comment by stephenrauch** [comment]
This is likely a duplicate of #65

**Comment by gbiddison** [comment]
It is related in that I think they both arise from bad packets, but the read_name method does it's own data unpacking presumably because the name is variable length, where the remaining data is unpacked directly by struct.unpack.

A struct.error handler around DNSIncoming.init would handle #65, and an IndexError handler would handle this one.  A generic handler handles both.

**Comment by stephenrauch** [comment]
@gbiddison 

I believe I have addressed this issue via #68.  If you could test it in your environment and report back that would be greatly appreciated.

**Comment by gbiddison** [comment]
:shipit: looks great!
